File: C++/longest-unequal-adjacent-groups-subsequence-ii.cpp
// Time:  O(n^2)
// Space: O(n)

// dp, backtracing
class Solution {
public:
    vector<string> getWordsInLongestSubsequence(int n, vector<string>& words, vector<int>& groups) {
        const auto& check = [&](int i, int j) {
            if (!(groups[i] != groups[j] && size(words[i]) == size(words[j]))) {
                return false;
            }
            int cnt = 0;
            for (int k = 0; k < size(words[i]); ++k) {
                cnt += words[i][k] != words[j][k] ? 1 : 0;
                if (cnt == 2) {
                    return false;
                }
            }
            return cnt == 1;
        };

        vector<pair<int, int>> dp(n, pair(1, -1));
        int i = -1;
        for (int i = n - 1; i >= 0; --i) {
            for (int j = i + 1; j < n; ++j) {
                if (check(j, i)) {
                    dp[i] = max(dp[i], pair(dp[j].first + 1, j));
                }
            }
        }
        vector<string> result;
        for (int i = distance(cbegin(dp), max_element(cbegin(dp), cend(dp)));i != -1; i = dp[i].second) {
            result.emplace_back(words[i]);
        }
        return result;
    }
};

// Time:  O(n^2)
// Space: O(n)
// dp, backtracing
class Solution2 {
public:
    vector<string> getWordsInLongestSubsequence(int n, vector<string>& words, vector<int>& groups) {
        const auto& check = [&](int i, int j) {
            if (!(groups[i] != groups[j] && size(words[i]) == size(words[j]))) {
                return false;
            }
            int cnt = 0;
            for (int k = 0; k < size(words[i]); ++k) {
                cnt += words[i][k] != words[j][k] ? 1 : 0;
                if (cnt == 2) {
                    return false;
                }
            }
            return cnt == 1;
        };

        vector<pair<int, int>> dp(n, pair(1, -1));
        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < i; ++j) {
                if (check(j, i)) {
                    dp[i] = max(dp[i], pair(dp[j].first + 1, j));
                }
            }
        }
        vector<string> result;
        for (int i = distance(cbegin(dp), max_element(cbegin(dp), cend(dp)));i != -1; i = dp[i].second) {
            result.emplace_back(words[i]);
        }
        reverse(begin(result), end(result));
        return result;
    }
};

File: Python/longest-unequal-adjacent-groups-subsequence-ii.py
# Time:  O(n^2)
# Space: O(n)

import itertools


# dp, backtracing
class Solution(object):
    def getWordsInLongestSubsequence(self, n, words, groups):
        """
        :type n: int
        :type words: List[str]
        :type groups: List[int]
        :rtype: List[str]
        """
        def check(s1, s2):
            return len(s1) == len(s2) and sum(a != b for a, b in itertools.izip(s1, s2)) == 1

        dp = [[1, -1] for _ in xrange(n)]
        for i in reversed(xrange(n)):
            for j in xrange(i+1, n):
                if groups[i] != groups[j] and check(words[j], words[i]):
                    dp[i] = max(dp[i], [dp[j][0]+1, j])
        result = []
        i = max(xrange(n), key=lambda x: dp[x])
        while i != -1:
            result.append(words[i])
            i = dp[i][1]
        return result


# Time:  O(n^2)
# Space: O(n)
import itertools


# dp, backtracing
class Solution2(object):
    def getWordsInLongestSubsequence(self, n, words, groups):
        """
        :type n: int
        :type words: List[str]
        :type groups: List[int]
        :rtype: List[str]
        """
        def check(s1, s2):
            return len(s1) == len(s2) and sum(a != b for a, b in itertools.izip(s1, s2)) == 1

        dp = [[1, -1] for _ in xrange(n)]
        for i in xrange(n):
            for j in xrange(i):
                if groups[i] != groups[j] and check(words[j], words[i]):
                    dp[i] = max(dp[i], [dp[j][0]+1, j])
        result = []
        i = max(xrange(n), key=lambda x: dp[x])
        while i != -1:
            result.append(words[i])
            i = dp[i][1]
        result.reverse()
        return result


# Time:  O(n^2)
# Space: O(n^2)
import itertools


# lis dp
class Solution3(object):
    def getWordsInLongestSubsequence(self, n, words, groups):
        """
        :type n: int
        :type words: List[str]
        :type groups: List[int]
        :rtype: List[str]
        """
        def check(s1, s2):
            return len(s1) == len(s2) and sum(a != b for a, b in itertools.izip(s1, s2)) == 1

        dp = [[] for _ in xrange(n)]
        for i in xrange(n):
            for j in xrange(i):
                if groups[i] != groups[j] and check(words[j], words[i]) and len(dp[j]) > len(dp[i]):
                    dp[i] = dp[j]
            dp[i] = dp[i]+[i]
        return map(lambda x: words[x], max(dp, key=lambda x: len(x)))
